File: day01/main.py
with open("input.txt") as file:
	inpt = list(map(int, file.read().splitlines()))

def increases(ints):
  result = 0
  for i in range(len(ints))[1:]:
    if (ints[i-1] < ints[i]):
      result += 1
  return result

def windowedIncreases(ints, k):
  result = 0
  for i in range(len(ints)-k):
    # To check whether a_i + ... + a_{i+k-1} < a_{i+1} + ... + a_{i+k}
    # we only need to check if a_i < a_{i+k}
    if (ints[i] < ints[i+k]):
      result += 1
  return result

print(f'Part 1: {increases(inpt)}')
print(f'Part 2: {windowedIncreases(inpt, 3)}')

File: day01/main.hs
main :: IO ()
main = do 
  ints <- parseFile
  putStrLn $ "Part 1: " <> show (increases ints)
  putStrLn $ "Part 2: " <> show (increases (map sum (windowed 3 ints))) 

increases :: Ord a => [a] -> Int
increases xs = foldr (\(x, y) acc -> acc + if x < y then 1 else 0) 0 $ zip xs (tail xs)

windowed :: Int -> [a] -> [[a]]
windowed k xs = [ take k ys | i <- [0..length xs - k], let ys = drop i xs ]

parseFile :: IO [Int]
parseFile = map read . lines <$> readFile "input.txt"
